diff --git a/include/game_unit.hpp b/include/game_unit.hpp
index 6346403..1bea901 100644
--- a/include/game_unit.hpp
+++ b/include/game_unit.hpp
@@ -1,3 +1,5 @@
+#pragma once
+
 #include "unit.hpp"
 #include "subject.hpp"
 
diff --git a/src/battle_monitor.cpp b/src/battle_monitor.cpp
index 7d675f7..6b53374 100644
--- a/src/battle_monitor.cpp
+++ b/src/battle_monitor.cpp
@@ -1,8 +1,13 @@
 #include <iostream>
 #include "battle_monitor.hpp"
 
+
 void BattleMonitor::do_update(void *event)
 {
+    // cast the downcasted void pointer to a UnitEvent pointer 
+    // so that we can access the message and print it out
     UnitEvent *e = static_cast<UnitEvent *>(event);
     std::cout << e->message << std::endl;
+    // also add the message to the messages vector so that we can later check it in tests
+    this->messages.push_back(e->message);
 }
\ No newline at end of file
diff --git a/src/revolver.cpp b/src/revolver.cpp
index 498fc73..fd599f8 100644
--- a/src/revolver.cpp
+++ b/src/revolver.cpp
@@ -1,7 +1,8 @@
 #include "revolver.hpp"
 
-Revolver::Revolver() : projectile(0)
+Revolver::Revolver() : projectile(1)
 {
+    // simple implementation just have the constructor intialize the projectile with a damage of 1
 }
 
 Projectile &Revolver::get_projectile()
diff --git a/src/soldier.cpp b/src/soldier.cpp
index 019eb51..4c440a2 100644
--- a/src/soldier.cpp
+++ b/src/soldier.cpp
@@ -9,8 +9,18 @@ Soldier::Soldier(std::string name, Weapon *weapon) : weapon(weapon), name(name)
 
 void Soldier::shootAt(Unit &unit)
 {
+    // uses the weapon associated with the class (which is made to be associated in construction)
+    // this weapons shootat function is then used which works by having the unit shot at
+    // use the takehit function
+	this->weapon->shootAt(unit);
 }
 
 void Soldier::takeHit(Projectile &projectile)
 {
+    // takes the hits from the shootat function it just announces damage taken (a more advance version would have hp)
+    // the projectile is gotten through get_projectile() which is an abstract class that we in this case
+    // implement in the revolver class
+	std::stringstream ss;
+	ss << this->name << " got hit for " << projectile.damage << " damage";
+	this->notify(UnitEvent{ss.str()});
 }
\ No newline at end of file
diff --git a/src/unit_selection.cpp b/src/unit_selection.cpp
index 4a472a3..16f20d3 100644
--- a/src/unit_selection.cpp
+++ b/src/unit_selection.cpp
@@ -3,18 +3,27 @@
 
 void UnitSelection::addToSelection(Unit &unit)
 {
+    // add unit pointer to selected vector
+    this->selected.push_back(&unit);
 }
 
+
 void UnitSelection::shootAt(Unit &unit)
 {
+    // loop through selected vector and have each unit use 
+    // the shootAt function on the unit passed to the function
     for (auto &u : this->selected)
     {
+        u->shootAt(unit);
     }
 }
 
-void UnitSelection::takeHit(Projectile &Projectile)
+void UnitSelection::takeHit(Projectile &projectile)
 {
+    // loops through selected vector and use takehit for each 
+    // unit in regard to the projectile passed
     for (auto &u : this->selected)
     {
+        u->takeHit(projectile);
     }
 }
